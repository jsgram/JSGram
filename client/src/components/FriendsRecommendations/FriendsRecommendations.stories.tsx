import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { MemoryRouter } from 'react-router';
import {FriendsRecomendations} from './index';

export const FriendsRecomendationsState = {
    loggedUsername: 'Maks',
    friendsRecommendations: {
        loading: false,
        users: [
            {
                _id: '1f',
                username: 'Diana',
                photoPath: 'https://jsgram-post-images.s3.eu-west-2.amazonaws.com/1570605706200',
            },
            {
                _id: '2f',
                username: 'Antony',
                photoPath: 'https://jsgram-post-images.s3.eu-west-2.amazonaws.com/1570605728486',
            },
            {
                _id: '3f',
                username: 'Bob',
                photoPath: 'https://jsgram-post-images.s3.eu-west-2.amazonaws.com/1570603193496',
            },
        ],
    },
};

const FriendsRecomendationsActions = {
    followUser: action('followUser'),
};

storiesOf('Friends Recomendations', module)
    .addDecorator((story: any): any => <MemoryRouter>{story()}</MemoryRouter>)
    .add('friends Recomendations', () =>
        <FriendsRecomendations {...{...FriendsRecomendationsState, ...FriendsRecomendationsActions}}/>);
