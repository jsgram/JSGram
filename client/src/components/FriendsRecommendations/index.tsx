import React from 'react';
import { Link } from 'react-router-dom';
import { Button } from 'reactstrap';
import noAvatar from '../../assets/noAvatar.png';

export interface IUser {
    photoPath: string;
    _id: string;
    username: string;
}

export interface IFriendsRecommendation {
    users: IUser[];
    loading: boolean;
}

interface IProps {
    loggedUsername: string;
    friendsRecommendations: IFriendsRecommendation;
    followUser: (body: {_id: string}) => void;
}

export const FriendsRecomendations = ({loggedUsername, friendsRecommendations, followUser}: IProps): JSX.Element => {
    const USERS_PER_PAGE = 4;
    const topRecommendations = friendsRecommendations.users.slice(0, USERS_PER_PAGE);

    return (
        <div className='mt-5'>
            <h4 className='text-center'>Suggestions for you:</h4>
            {topRecommendations.map((user: IUser) => (
                        <div className='d-flex justify-content-between mb-2' key={user._id}>
                            <div className='wrapper'>
                            <Link to={`/profile/${user.username}`}>
                                <img
                            src={user.photoPath || noAvatar}
                            alt='avatar'
                            width={64}
                            height={64}
                            className='img-fluid rounded-circle'
                        /></Link>
                        <Link to={`/profile/${user.username}`} className='mt-1 ml-3 mr-4
                                  text-dark'>{user.username}</Link>
                        </div>
                        <Button className='align-self-center' color='danger'
                                onClick={(e: any): void => {
                                    followUser({_id: user._id});
                                    e.target.disabled = true;
                                }}>Follow</Button>
                        </div>
                ),
            )}
            {friendsRecommendations.users.length > topRecommendations.length &&
                <Link to={`/profile/${loggedUsername}/recommendations`} className='mr-2 text-decoration-none'>
                        See all recommendations</Link>}
      </div>
    );
};
