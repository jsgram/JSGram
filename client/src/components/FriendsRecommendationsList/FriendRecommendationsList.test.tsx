import React from 'react';
import FriendsRecommendations, { FriendsRecommendationsList } from './index';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

describe('FriendsRecommendationsList smart component', () => {
    let renderer: any;
    let store: any;
    const props = {
        friendsRecommendations: {
            users: [],
            loading: true,
        },
        getRecommendations: jest.fn(() => 'some value'),
        followUser: jest.fn(() => 'some value'),
    };
    const initialState = {
        newsFeed: {
            friendsRecommendations: {
                users: [],
                loading: true,
            },
        },
        profileEdit: {
            newUsername: 'some value',
        },
        feed: {
            loggedUsername: 'some value',
        },
        search: {
            searchResults: [],
        },
    };

    beforeEach(() => {
        store = configureStore()(initialState);
        renderer = shallow(<BrowserRouter><FriendsRecommendationsList {...props}/></BrowserRouter>);
    });

    test('componentDidMount - success', () => {
        renderer.instance().componentDidMount();
        expect(props.getRecommendations).not.toBeCalled();
    });

    test('render - success', () => {
        renderer = shallow(<Provider store={store}>
            <BrowserRouter><FriendsRecommendations {...props}/></BrowserRouter>
        </Provider>);

        expect(renderer).toMatchSnapshot();
    });
});
