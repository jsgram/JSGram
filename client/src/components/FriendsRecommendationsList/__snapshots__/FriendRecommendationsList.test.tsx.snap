// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendsRecommendationsList smart component render - success 1`] = `ShallowWrapper {}`;
